### ISSUE TITLE
[tune] Trial workers are not co-located

### ISSUE THREAD

**Issue by roireshef** [ray-project/ray#12360]
### What is the problem?
Using Tune as a scheduler and resource manager for multiple trials - especially where each trial can fit into a single Ray node (machine) - the expected behaviour is that Tune will schedule all workers for a trial on the same node. In reality, it seems like there's mix & match between nodes and RolloutWorker processes from different trials.

*Ray version and other system information (Python version, TensorFlow version, OS):*
Ray 0.7.2
Docker 19.0.3
PyTorch 1.5.1
Ubuntu 18.04.5

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

Run a Tune experiment with multiple Ray nodes (I used 10 nodes), with the following config:
```
    config = DEFAULT_CONFIG
    config.update({
        "env": <your own environment of choice>,
        "use_pytorch": True,
        "num_workers": 31,
        "num_envs_per_worker": tune.grid_search([3, 5]),
   })

    tune.run(
        A3CTrainer,
        name=<experiment name>,
        stop={
            "timesteps_total": 200e6,  # Million steps
        },
        num_samples=1,
        config=config,
        checkpoint_freq=50,
        max_failures=-1
    )
```
If we cannot run your script, we cannot fix your issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by roireshef** [comment]
Just to provide a better evidence, this is a snapshot of Tune's periodic summary table:
_== Status ==
Using FIFO scheduling algorithm.
Resources requested: 320/320 CPUs, 0/0 GPUs
Memory usage on this node: 33.4/201.4 GB
Result logdir: /home/kz430x/ray_results/old_everything_debug
Number of trials: 48 ({'RUNNING': 10, 'PENDING': 38})
PENDING trials:
 - new_base_NegotiationMergeHandoffFullLateralEnv_10_JERK_COEF=0.01,KNOT_JERK_DIFF_COEF=0.1,architecture=AttentionNetWithLastAction_V2,num_envs_per_worker=5,sample_batch_size=100:	PENDING
 - new_base_NegotiationMergeHandoffFullLateralEnv_11_JERK_COEF=0.1,KNOT_JERK_DIFF_COEF=0.1,architecture=AttentionNetWithLastAction_V2,num_envs_per_worker=5,sample_batch_size=100:	PENDING
 - new_base_NegotiationMergeHandoffFullLateralEnv_12_JERK_COEF=0.01,KNOT_JERK_DIFF_COEF=0.01,architecture=AttentionNetWithLastAction_V2_Tanh,num_envs_per_worker=5,sample_batch_size=100:	PENDING
 - new_base_NegotiationMergeHandoffFullLateralEnv_13_JERK_COEF=0.1,KNOT_JERK_DIFF_COEF=0.01,architecture=AttentionNetWithLastAction_V2_Tanh,num_envs_per_worker=5,sample_batch_size=100:	PENDING
 - new_base_NegotiationMergeHandoffFullLateralEnv_14_JERK_COEF=0.01,KNOT_JERK_DIFF_COEF=0.1,architecture=AttentionNetWithLastAction_V2_Tanh,num_envs_per_worker=5,sample_batch_size=100:	PENDING
  ... 28 not shown
 - new_base_NegotiationMergeHandoffFullLateralEnv_43_JERK_COEF=0.1,KNOT_JERK_DIFF_COEF=0.1,architecture=AttentionNetWithLastAction_V2,num_envs_per_worker=5,sample_batch_size=300:	PENDING
 - new_base_NegotiationMergeHandoffFullLateralEnv_44_JERK_COEF=0.01,KNOT_JERK_DIFF_COEF=0.01,architecture=AttentionNetWithLastAction_V2_Tanh,num_envs_per_worker=5,sample_batch_size=300:	PENDING
 - new_base_NegotiationMergeHandoffFullLateralEnv_45_JERK_COEF=0.1,KNOT_JERK_DIFF_COEF=0.01,architecture=AttentionNetWithLastAction_V2_Tanh,num_envs_per_worker=5,sample_batch_size=300:	PENDING
 - new_base_NegotiationMergeHandoffFullLateralEnv_46_JERK_COEF=0.01,KNOT_JERK_DIFF_COEF=0.1,architecture=AttentionNetWithLastAction_V2_Tanh,num_envs_per_worker=5,sample_batch_size=300:	PENDING
 - new_base_NegotiationMergeHandoffFullLateralEnv_47_JERK_COEF=0.1,KNOT_JERK_DIFF_COEF=0.1,architecture=AttentionNetWithLastAction_V2_Tanh,num_envs_per_worker=5,sample_batch_size=300:	PENDING
RUNNING trials:
 - new_base_NegotiationMergeHandoffFullLateralEnv_0_JERK_COEF=0.01,KNOT_JERK_DIFF_COEF=0.01,architecture=AttentionNetWithLastAction_V2,num_envs_per_worker=3,sample_batch_size=100:	RUNNING, [32 CPUs, 0 GPUs], [**dcwipphpc1854.edc.nam.gm.com** pid=70], 3187 s, 45 iter, 1411885 ts, 0.885 rew
 - new_base_NegotiationMergeHandoffFullLateralEnv_1_JERK_COEF=0.1,KNOT_JERK_DIFF_COEF=0.01,architecture=AttentionNetWithLastAction_V2,num_envs_per_worker=3,sample_batch_size=100:	RUNNING, [32 CPUs, 0 GPUs], [**dcwipphpc1853.edc.nam.gm.com** pid=41], 3179 s, 43 iter, 1341466 ts, 0.786 rew
 - new_base_NegotiationMergeHandoffFullLateralEnv_2_JERK_COEF=0.01,KNOT_JERK_DIFF_COEF=0.1,architecture=AttentionNetWithLastAction_V2,num_envs_per_worker=3,sample_batch_size=100:	RUNNING, [32 CPUs, 0 GPUs], [**dcwipphpc1854.edc.nam.gm.com** pid=69], 3201 s, 45 iter, 1405764 ts, 0.754 rew
 - new_base_NegotiationMergeHandoffFullLateralEnv_3_JERK_COEF=0.1,KNOT_JERK_DIFF_COEF=0.1,architecture=AttentionNetWithLastAction_V2,num_envs_per_worker=3,sample_batch_size=100:	RUNNING, [32 CPUs, 0 GPUs], [pid=56], 3164 s, 46 iter, 1436004 ts, 0.695 rew
 - new_base_NegotiationMergeHandoffFullLateralEnv_4_JERK_COEF=0.01,KNOT_JERK_DIFF_COEF=0.01,architecture=AttentionNetWithLastAction_V2_Tanh,num_envs_per_worker=3,sample_batch_size=100:	RUNNING, [32 CPUs, 0 GPUs], [dcwipphpc1832.edc.nam.gm.com pid=57], 3222 s, 48 iter, 1516554 ts, 0.825 rew
 - new_base_NegotiationMergeHandoffFullLateralEnv_5_JERK_COEF=0.1,KNOT_JERK_DIFF_COEF=0.01,architecture=AttentionNetWithLastAction_V2_Tanh,num_envs_per_worker=3,sample_batch_size=100:	RUNNING, [32 CPUs, 0 GPUs], [pid=70], 3184 s, 46 iter, 1444697 ts, 0.718 rew
 - new_base_NegotiationMergeHandoffFullLateralEnv_6_JERK_COEF=0.01,KNOT_JERK_DIFF_COEF=0.1,architecture=AttentionNetWithLastAction_V2_Tanh,num_envs_per_worker=3,sample_batch_size=100:	RUNNING, [32 CPUs, 0 GPUs], [dcwipphpc1878.edc.nam.gm.com pid=63], 3214 s, 47 iter, 1477262 ts, 0.749 rew
 - new_base_NegotiationMergeHandoffFullLateralEnv_7_JERK_COEF=0.1,KNOT_JERK_DIFF_COEF=0.1,architecture=AttentionNetWithLastAction_V2_Tanh,num_envs_per_worker=3,sample_batch_size=100:	RUNNING, [32 CPUs, 0 GPUs], [**dcwipphpc1853.edc.nam.gm.com** pid=64], 3173 s, 45 iter, 1405049 ts, 0.69 rew
 - new_base_NegotiationMergeHandoffFullLateralEnv_8_JERK_COEF=0.01,KNOT_JERK_DIFF_COEF=0.01,architecture=AttentionNetWithLastAction_V2,num_envs_per_worker=5,sample_batch_size=100:	RUNNING, [32 CPUs, 0 GPUs], [dcwipphpc1904.edc.nam.gm.com pid=53], 3148 s, 28 iter, 1441463 ts, 0.861 rew
 - new_base_NegotiationMergeHandoffFullLateralEnv_9_JERK_COEF=0.1,KNOT_JERK_DIFF_COEF=0.01,architecture=AttentionNetWithLastAction_V2,num_envs_per_worker=5,sample_batch_size=100:	RUNNING, [32 CPUs, 0 GPUs], [dcwipphpc1878.edc.nam.gm.com pid=52], 3132 s, 26 iter, 1332728 ts, 0.649 rew_

Note how at least two trials (the ones that are in bold) are sharing the same host-address (the same Ray node). 
Running _top_ on one of those nodes provide the following output:
![tune_issue](https://user-images.githubusercontent.com/28055761/100140698-fcbc3980-2e99-11eb-8183-48abccfbb3a1.png)

See how some ray processes have 3 environments and some have 5 (_num_envs_per_worker_ is a HP with values 3,5). This is on the same machine, so obviously those are RolloutWorker processes from different trials.

**Comment by richardliaw** [comment]
@sven1977 this isn't a Tune issue; this is something you need to implement in the rllib trainer (using placement groups!)

cc @rkooo567 who can help you out here.

**Comment by sven1977** [comment]
@richardliaw Ah, sorry, I misunderstood. Thanks for the hint!

**Comment by roireshef** [comment]
I just wanted to point out that the current state is that Tune can't be trusted to execute multiple trials on a Ray cluster and provide valid results because those trials that have workers spread across different machines suffer from higher latency in the communication of driver <-> workers, creating stale (off-policy) gradients/samples. This eventually messes up the results of the hyperparameter tuning. Therefor, this sounds to me more like a bug rather than an enhancement, as it makes "Tune: Scalable Hyperparameter Tuning" **unreliable for doing hyperparameter tuning when running more than a single trial at once**.

**Comment by roireshef** [comment]
Hey @richardliaw I just read https://docs.ray.io/en/master/placement-group.html?highlight=placement%20groups

Automating the creation of placement groups in the RLLib use case depends on the config provided (specifically _num_workers_ and _num_gpus_) as well as the hardware setup - you'd generally like to schedule every trial on a single node, but if it has num_workers > node-cpus then you have the distribute training across nodes. I think STRICT_SPREAD makes the most sense here, right?

That two types of information appear to be exposed to Tune, rather than the Trainer/Trainable classes, isn't it? Would you suggest implementing it anywhere else?

Also, how long back is the placement group API supported? (from what version is it stable)

**Comment by richardliaw** [comment]
All of this information is available to both Tune and the Trainer.

This must be implemented at Trainer level because the Trainer is the part
of the stack that launches rollout workers. Not Tune.

Stability for placement groups is recent (1.0).

**Comment by roireshef** [comment]
Hi @richardliaw, actually I'm experiencing very noticeable degradation when using Tune on clusters of more than, say 1-2 nodes/trials that are running simultaneously in a Tune experiment. I don't remember having those same slowdowns or reliability-issues when using Tune with 10-15 trials/nodes (1 trial per node) in _Ray 0.7.2_ , and I was doing that **a lot**. 

Has anything changed in Tune's default behaviour (with placement groups not implemented)? I mean, I understand the best practice would be to just implement it, but I'd like to figure out if there's some fast solution for the meantime until it's implemented.

Thanks.

**Comment by richardliaw** [comment]
Hmm, you're on the latest master?

**Comment by roireshef** [comment]
> Hmm, you're on the latest master?

@richardliaw I'm not, because of https://github.com/ray-project/ray/issues/12543. But I did pull latest nightly wheel yesterday and that's what I'm working with. Any news since yesterday...?

**Comment by richardliaw** [comment]
@sven1977 would it make sense to prioritize placementgroup implementation for A3C just so Roi can try things out?

I can send you a simple design doc that we used in our projects.

**Comment by roireshef** [comment]
@richardliaw I wouldn't like you guys to re-prioritize your task based on my personal preferences... The reason I'm pushing this issue is that I believe there are two (or more) issues here - while I understand the best practice is to implement the placementgroup thing, I don't understand why not having them implemented creates a memory overload for clusters of more than a handful of nodes. Seems like something other than the requirement of placementgroups has been broken along the way, don't you think? Isn't it worth debugging this to understand what's the cause for the memory issues, or do you believe the fact workers are not co-located can alone create all of this mess...?

**Comment by richardliaw** [comment]
Actually, yeah I think something bad is broken/deadlocks. 

This person here reports a 4x slowdown over time: https://ray-distributed.slack.com/archives/CN2RGCHRR/p1606844191274400?thread_ts=1606642434.261900&cid=CN2RGCHRR

**Comment by rkooo567** [comment]
Btw, there's a high possibility this is the root cause; https://github.com/ray-project/ray/issues/12546

So GCS actor management basically distributes actors rather than colocating them when it is possible. This also explains why there was object broadcast issues coming up. Since actors are more distributed, there were more object transfers. 

@ffbin will work on this issue shortly, and let's see if this can relieve the issue;

**Comment by roireshef** [comment]
@rkooo567, @ffbin - Thanks! I'll be following this thread.

**Comment by roireshef** [comment]
This is probably related to https://github.com/ray-project/ray/issues/12352 as well. I hope the solution will it as well.

**Comment by roireshef** [comment]
@rkooo567 Did you end up testing if those slowdowns are fixed? What version is this fix going into?

**Comment by rkooo567** [comment]
Yeah the fix has been pushed to the master branch, so if you can verify if it fixes your issue, that'll be awesome! 

https://github.com/ray-project/ray/pull/12707

**Comment by roireshef** [comment]
@rkooo567 I'm having difficulties doing that, since the soft-links-based setup still doesn't work (https://github.com/ray-project/ray/issues/12543). How can I tell when was the last time the "nightly" build was built? Is it actually building **every night** out of the master branch? If I used the latest nightly build - am I guaranteed to have this fix inside? Also, assuming I'm using a nightly build from 2 weeks ago - should I just run `ray install-nightly` to update to latest build?

**Comment by rkooo567** [comment]
It is built every commit. So yes. The nightly will contain the latest changes

**Comment by roireshef** [comment]
OK, This actually works quite well now, running Tune with 27 trials (1 node x 32 cores each).
Thanks a lot for your support!!

**Comment by richardliaw** [comment]
wow! great!
